### ISSUE TITLE
MultiDex for debug builds not properly initialized below API 20

### ISSUE THREAD

**Issue by ialokim** [grote/Transportr#715]
**To Reproduce**
1. Open Transportr app (current master) on API level 16
2. Get following stacktrace:
```
W/dalvikvm: VFY: unable to find class referenced in signature (Lcom/mapbox/services/android/telemetry/location/LocationEngine;)
E/dalvikvm: Could not find class 'com.mapbox.services.android.telemetry.location.LocationEngineProvider', referenced from method com.mapbox.mapboxsdk.Mapbox.getInstance
W/dalvikvm: VFY: unable to resolve new-instance 410 (Lcom/mapbox/services/android/telemetry/location/LocationEngineProvider;) in Lcom/mapbox/mapboxsdk/Mapbox;
D/dalvikvm: VFY: replacing opcode 0x22 at 0x000b
D/dalvikvm: GC_CONCURRENT freed 158K, 3% free 10981K/11271K, paused 11ms+10ms, total 29ms
W/dalvikvm: VFY: unable to find class referenced in signature (Lcom/mapbox/services/android/telemetry/location/LocationEngine;)
E/dalvikvm: Could not find class 'com.mapbox.mapboxsdk.EmptyLocationSource', referenced from method com.mapbox.mapboxsdk.Mapbox.getLocationSource
W/dalvikvm: VFY: unable to resolve new-instance 330 (Lcom/mapbox/mapboxsdk/EmptyLocationSource;) in Lcom/mapbox/mapboxsdk/Mapbox;
D/dalvikvm: VFY: replacing opcode 0x22 at 0x0000
W/dalvikvm: VFY: unable to resolve static field 156 (MAPBOX_LOCALE) in Lcom/mapbox/mapboxsdk/constants/MapboxConstants;
D/dalvikvm: VFY: replacing opcode 0x62 at 0x000a
E/dalvikvm: Could not find class 'com.mapbox.mapboxsdk.exceptions.MapboxConfigurationException', referenced from method com.mapbox.mapboxsdk.Mapbox.validateAccessToken
W/dalvikvm: VFY: unable to resolve new-instance 344 (Lcom/mapbox/mapboxsdk/exceptions/MapboxConfigurationException;) in Lcom/mapbox/mapboxsdk/Mapbox;
D/dalvikvm: VFY: replacing opcode 0x22 at 0x0027
E/dalvikvm: Could not find class 'com.mapbox.mapboxsdk.exceptions.MapboxConfigurationException', referenced from method com.mapbox.mapboxsdk.Mapbox.validateMapbox
W/dalvikvm: VFY: unable to resolve new-instance 344 (Lcom/mapbox/mapboxsdk/exceptions/MapboxConfigurationException;) in Lcom/mapbox/mapboxsdk/Mapbox;
D/dalvikvm: VFY: replacing opcode 0x22 at 0x0005
    DexOpt: unable to opt direct call 0x01f3 at 0x0d in Lcom/mapbox/mapboxsdk/Mapbox;.getInstance
    DexOpt: unable to opt direct call 0x013b at 0x02 in Lcom/mapbox/mapboxsdk/Mapbox;.getLocationSource
    DexOpt: unable to opt direct call 0x0146 at 0x29 in Lcom/mapbox/mapboxsdk/Mapbox;.validateAccessToken
    DexOpt: unable to opt direct call 0x0146 at 0x07 in Lcom/mapbox/mapboxsdk/Mapbox;.validateMapbox
D/AndroidRuntime: Shutting down VM
W/dalvikvm: threadid=1: thread exiting with uncaught exception (group=0xa669c228)
E/AndroidRuntime: FATAL EXCEPTION: main
    java.lang.NoClassDefFoundError: com.mapbox.services.android.telemetry.location.LocationEngineProvider
        at com.mapbox.mapboxsdk.Mapbox.getInstance(Mapbox.java:54)
        at de.grobox.transportr.TransportrApplication.onCreate(TransportrApplication.java:42)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:999)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4151)
        at android.app.ActivityThread.access$1300(ActivityThread.java:130)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1255)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:137)
        at android.app.ActivityThread.main(ActivityThread.java:4745)
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:511)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
        at dalvik.system.NativeStart.main(Native Method)
```

**Versions (please complete the following information):**
 - **Transportr Version:** 2.1.0
 - **Device:** Emulator
 - **Android Version:** 4.1

**Comment by grote** [comment]
maybe proguard optimizing it out?

**Comment by ialokim** [comment]
Strangly this does not happen on higher API levels (tested at least with 22).

**Comment by grote** [comment]
strange indeed (so probably not proguard) and we didn't even upgrade mapbox, right?

**Comment by ialokim** [comment]
> we didn't even upgrade mapbox, right?

True, we should check though from which version on this error happens.

**Comment by ialokim** [comment]
Btw, would it be possible to run Android Instrumentation tests on different API levels in Travis CI to notice such kind of problems?

**Comment by grote** [comment]
Yeah, you could set up this https://github.com/grote/Transportr/blob/master/.travis.yml#L30 for different API levels. Maybe in different test tracks?

**Comment by ialokim** [comment]
This does not happen when building a release, but only for the debug build variant. May be indeed connected to Proguard (while other API levels do not crash for debug builds either).

**Comment by ialokim** [comment]
It seems to be related to [native multiDex support starting with API 20](https://developer.android.com/studio/build/multidex). It seems multiDex is indeed needed for debug builds (probably due to additional `androidTest` dependencies?). However, MultiDex is not properly initialized within https://github.com/grote/Transportr/blob/fe7036716f61f9d25e072261d5aafa868f207175/app/src/main/java/de/grobox/transportr/TransportrApplication.java#L35

As multiDex is only enabled for debug builds
https://github.com/grote/Transportr/blob/fe7036716f61f9d25e072261d5aafa868f207175/app/build.gradle#L43
this cannot be easily commented in again.

Do you know why this part has been commented out? Perhaps we could enable these multiDex related things only for API below 20 somehow?

**Comment by ialokim** [comment]
Following https://stackoverflow.com/a/55223491, I was able to enable and initialize multiDex only for debug builds. Is that what we want?

**Comment by grote** [comment]
Ideally, we can get rid of multidex completely, don't even remember why it was needed in the first place.

**Comment by ialokim** [comment]
Unfortunately it is needed for debug builds (perhaps due to not shrinking resources and/or additional `androidTestImplementations`) and for API level below 20. Starting with Android 5, Multidex is supported out of the box. It should not harm the release builds at all.

**Comment by ialokim** [comment]
Closed by #723.
